### DISCUSSION TITLE
npm should not forcibly replaces symlinked node_modules

### DISCUSSION THREAD

**Discussion by vicary** [npm/feedback#571] [CLI] [answered]
This behavior is introduced between npm 7.15 and 7.21.

The new `npm install` behavior is to always replace a symlinked node_modules with an actual directory, which breaks the usage of node_modules.nosync and a few existing npm modules that favors this behavior, e.g. nosync-icloud

If this breaking change is intended, it should be introduced with a major version upgrade, shouldn’t it?

**Comment by MylesBorins** [comment, marked as answer]
if you think the npm CLI has introduced an unexpected breaking change you should open an issue reporting a bug at https://github.com/npm/cli

**Reply by vicary** [reply]
Thanks for the reply, there is an existing issue tracking this one. https://github.com/npm/cli/issues/3669

**Reply by MylesBorins** [reply]
Thanks for sharing. in general there isn't anything that can be accomplished via this feedback process if there is an existing issue in another related npm repository

**Comment by trusktr** [comment]
This is not a new issue, this issue has plagued us since the beginning of (npm's) time. This is what I always have to currently do to mitigate the issue when not using a tool like Lerna:

- take note of all symlinked packages (f.e. with `find node_modules -type l | grep -v .bin`)
- remove all the links, otherwise not only will `npm` replace those, it will first clobber the `node_modules` in those linked packages (ouch).
- run `npm install`, which effectively replaces the links your own way, without clobbering
- based on the links you tracked, re-link everything.

It is really painful. But luckily there are tools that help with this. With Lerna, my process is:

- `rm -rf node_modules` in the project that has linked packages.
- run `npm install`
- go up to the root and run `lerna bootstrap` which automatically re-links everything

With the new npm 7 workspaces, using `npm install` might be an alternative solution to Lerna assuming you make use of a `workspaces` config in your package.json. So then the steps could be the following,

- `rm -rf node_modules` in the project that has linked packages.
- go up to the root and run `npm install` which automatically re-links everything

and maybe there's no need to delete `node_modules` with npm 7 if workspaces are configured, but I haven't verified.

I've reported this bug in multiple issue trackers that `npm` had before, but they have closed those systems and I haven't bothered to open yet another issue.

I think the team even claimed that npm 7 would finally get rid of this problem back in one of those trackers, and if they did, then this is a regression, but I haven't verified this either. I just use the above steps knowing it always works.

**Reply by vicary** [reply]
TBH Isaacs has a tendency to close issues outright since the early days of npm, leaving little rooms for reasoning. I have no hope for npm as a company. You have my appreciation for at least tried multiple occasions, also thanks for your workarounds!
